### PR TITLE
feat: Added documentation for UI unit test component query

### PR DESCRIPTION
**Pull request by mcollovati**
Part of vaadin/testbench#1338

### CODE REVIEW

**Comment by tarekoraby** [comment]
If more stuff needs to be added in the future for UI unit testing, I think we should consider splitting the whole thing into several sub-pages.

File: articles/tools/testbench/ui-unit-testing.asciidoc [hunk 1]
```diff
@@ -81,3 +81,61 @@ class MyViewTest extends UIUnitTest {
 
 }
 ----
+
+[[testbench.uiunit.search-components]]
+== Searching for Components
+
+[classname]`UIUnitTest` base class is able to get the instantiated view, but child components may not be always accessible directly.
+For example components may be stored in fields with private visibility or they may even not be referenced at all in the view class.
+To overcome this limitations, [classname]`UIUnitTest` provides a query functionality, that lets you scan the UI and filter for the components you need to interact with in test methods.
```
Comment (caalador) [inline articles/tools/testbench/ui-unit-testing.asciidoc:90]:
```suggestion
To overcome this limitation, [classname]`UIUnitTest` provides a Component query functionality, that lets you search the component tree for the components you need to interact with in test methods.
```

File: articles/tools/testbench/ui-unit-testing.asciidoc [hunk 2]
```diff
@@ -81,3 +81,61 @@ class MyViewTest extends UIUnitTest {
...
+Query object can be obtained by calling the [methodname]`select()` method, providing the type of the component you are searching for.
```
Comment (caalador) [inline articles/tools/testbench/ui-unit-testing.asciidoc:91]:
```suggestion
The `ComponentQuery` object can be obtained by calling the [methodname]`select()` method, providing the type of the component you are searching for.
```

**Comment by mcollovati** [comment]
> If more stuff needs to be added in the future for UI unit testing, I think we should consider splitting the whole thing into several sub-pages.

For sure there will be more docs on the ongoing weeks, and I think also the section related to this PR will grow.
Perhaps I can move it to another page already in this PR

**Comment by mcollovati** [comment]
If I correctly understand I should create a new `ui-unit-testing` folder, move the current content into `index.asciidoc` and then create the new page for `Searching components section`. Is this correct? Can i do all in this PR o should I first open another one to create the new folder?

**Comment by tarekoraby** [comment]
It can be done in this PR if you want. I suggest creating a `ui-unit-testing` folder with a minimal `index.adoc`:
```
---
title: UI Unit Testing
order: 160
---
```

Then create three adoc files inside the folder: 

- `overview.adoc` that contains everything until the limitations section. 
- `getting-started.adoc` for the getting started section (as an aside, I wonder if this section is still missing some instruction on how to actually run the tests. I assume `mvn test` is all that's needed, but it's still worth spelling it out).
- `searching-components` for the Searching for Components section.

**Comment by tarekoraby** [comment]
@mcollovati @caalador as there are many PRs open now related to this page, it might be easier to do the restructuring separately in a new PR when everything is merged.

File: articles/tools/testbench/ui-unit-testing.asciidoc (hunk 1 above)
Comment (mcollovati) [reply articles/tools/testbench/ui-unit-testing.asciidoc]: done

File: articles/tools/testbench/ui-unit-testing.asciidoc (hunk 2 above)
Comment (mcollovati) [reply articles/tools/testbench/ui-unit-testing.asciidoc]: done

File: articles/tools/testbench/ui-unit-testing.asciidoc [hunk 3]
```diff
@@ -81,3 +81,62 @@ class MyViewTest extends UIUnitTest {
 
 }
 ----
+
+[[testbench.uiunit.search-components]]
+== Searching for Components
+
+[classname]`UIUnitTest` base class is able to get the instantiated view, but child components may not be always accessible directly.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing.asciidoc:88]:
```suggestion
The [classname]`UIUnitTest` base class is able to get the instantiated view, but child components may not be always accessible directly.
```

File: articles/tools/testbench/ui-unit-testing.asciidoc [hunk 4]
```diff
@@ -81,3 +81,62 @@ class MyViewTest extends UIUnitTest {
...
+For example components may be stored in fields with private visibility or they may even not be referenced at all in the view class.
+To overcome this limitation, [classname]`UIUnitTest` provides a Component query functionality, that lets you search the component tree for the components you need to interact with in test methods.
+
+The [classname]`ComponentQuery` object can be obtained by calling the [methodname]`$()` method, providing the type of the component you are searching for.
+You can also restrict search scope to the children of current view, using the [methodname]`$view()` method.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing.asciidoc:93]:
```suggestion
You can also restrict search scope to the children of the current view, using the [methodname]`$view()` method.
```

File: articles/tools/testbench/ui-unit-testing.asciidoc [hunk 5]
```diff
@@ -81,3 +81,62 @@ class MyViewTest extends UIUnitTest {
...
+The query object has many filtering utilities that can be used to refine the search.
+For example you can filter by component `id`, by a property value, or using custom predicates on potential candidates.
+Once the query is ready with all required conditions configured, you can get test wrappers for specific components found by query execution by calling terminal operators.
+Example of terminal operators are [methodname]`first()`, [methodname]`last()`, [methodname]`atIndex()`, [methodname]`all()` and [methodname]`id()`.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing.asciidoc:97]:
```suggestion
Examples of terminal operators are [methodname]`first()`, [methodname]`last()`, [methodname]`atIndex()`, [methodname]`all()` and [methodname]`id()`.
```

File: articles/tools/testbench/ui-unit-testing.asciidoc [hunk 6]
```diff
@@ -81,3 +81,62 @@ class MyViewTest extends UIUnitTest {
...
+Predicate<TextField> fieldHasNotValue = field -> field.getOptionalValue().isEmpty();
+Predicate<TextField> fieldIsInvalid = TextField::isInvalid;
+TextFieldWrap<TextField, String> textField = $view(TextField.class)
+        .withCondition(fieldHasNotValue.or(fieldIsInvalid))
+        .first();
+
+----
+
+
+Sometimes you may need to query for components nested inside the UI, in a hierarchy composed by many different types of components.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing.asciidoc:123]:
```suggestion
Sometimes you may need to do a query for components nested inside the UI, in a hierarchy composed of many different types of components.
```

File: articles/tools/testbench/ui-unit-testing.asciidoc [hunk 7]
```diff
@@ -81,3 +81,62 @@ class MyViewTest extends UIUnitTest {
...
+To simplify such situations, the query object offers methods to chain a new query starting on a found component, so that complex query can be created in a fluent way.
+The [methodname]`thenOn()` method and its variants, for example [methodname]`thenOnFirst()`, provides you a new query object for the given component type, setting the search scope to the component selected from current query.
```
Comment (dougjewsbury) [inline articles/tools/testbench/ui-unit-testing.asciidoc:125]:
```suggestion
The [methodname]`thenOn()` method and its variants, for example [methodname]`thenOnFirst()`, provide you with a new query object for the given component type, setting the search scope to the component selected from the current query.
```

**Review by dougjewsbury — COMMENTED** [review]
AsciiDoc looks good.

**Review by caalador — APPROVED** [review]

**Pull request merged**